### PR TITLE
Introduced map mode for batch pipeline step...

### PR DESCRIPTION
**Pull request by kevinstyp**
..., which keeps separate Maps/Lists for each batchSeparationTag, which timeout independently of each other, allowing samples to change the tag inbetween the stream more often.

### CODE REVIEW

**Review by alek-thunder — COMMENTED** [review]
As we already discussed, I am still skeptical regarding this... We have a rather clear definition of what the default batch step should look like and how it should handle batches. From my point of view, this change is very specific and I do not see the necessity to implement it across all other language implementation. My proposed solution would be to create an extra class for this (MapBatchStep or something) that overrides writeSample, addBatchHandler and doClose

**Comment by alek-thunder** [comment]
Which is not possible since the batch step is an integral part of the bitflow script... The other, (in my opinion) better approach is to fork and, to prevent the memory leak through continuously changing tag values, implement an timeout after which subpiplines within the fork are pruned.
Maybe @antongulenko has further comments on this.

**Comment by antongulenko** [comment]
I have the same concern... The map mode basically completely ignores the rest of the batch step through if() conditions at the critical places. The only shared parameter is the batch separation tag, and it has a different meaning for the map mode. This is a definite indicator of bad design.

BUT big benefit of having it as a batch step in the bitflow script language is, that new processing steps can easily be written for this way of collecting/merging samples. I do see the possibility of reusing this, and the merging strategy based on tags is similar to some of the other merging steps that we wrote in the past.

My suggestion to clean the code and keep the functionality: Split the new BatchPipelineStep in 3 classes (yeah): abstract superclass, the normal batch step, and the new one. Then the abstract superclass will have the static method createFromParameters(), which is used by the script compiler to get a BatchPipelineStep instance. This method will create an instance of the appropriate subclass based on a flag-parameter passed from the script.
When splitting the BatchPipelineStep, the top prio is to avoid duplicated code :)

Also, please think of a better name than "mapMode". The name comes from the implementation strategy based on maps, but does not convey what the step actually does. I have no suggestions right now, but trust that you find something ;)

**Comment by antongulenko** [comment]
Lets keep this PR open for the proposed refactoring

**Comment by kevinstyp** [comment]
As the most common use case of the "map-mode" would be merging or collecting, i would suggest to rename it to "mergeMode" (other suggestions: "multiTimeout", "multiTags", "desynchedMode").
The current implementation has the advantage of not needing to add another keyword to the bitflow-script. As far as i understand, your suggestion should make it possible to not need another keyword but still being able to refactor the classes, based on the passed parameters. In addition to that, this refactoring should make it possible to add more possible batch-modes in the future easily.

File: src/main/java/bitflow4j/steps/AbstractBatchPipelineStep.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import bitflow4j.script.registry.RegisteredParameter;
+import bitflow4j.script.registry.RegisteredParameterList;
+import bitflow4j.task.TaskPool;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.logging.Logger;
+
+/**
+ * Instead of immediately handling every Sample, fill up a list of samples based on a number of criteria (time, tag values, ...).
```
Comment (antongulenko) [inline src/main/java/bitflow4j/steps/AbstractBatchPipelineStep.java]: Pls update class comment, e.g. the class is not final anymore

**Review by antongulenko — DISMISSED** [review]

**Review by antongulenko — APPROVED** [review]

**Pull request merged**